### PR TITLE
Made plugin forwards compatible

### PR DESCRIPTION
**Pull request by yannickl88**
Made plugin forwards compatible.

Also added a cache per project for version and command path so we do not have unnecessary calls every time the user hits save.

### CODE REVIEW

File: lib/atom-css-sniffer.js [hunk 1]
```diff
@@ -4,16 +4,96 @@ let helpers;
 let path;
 let filesystem;
 
+// Caches
+let versions = {};
+let cmds = {}
```
Comment (rickprent) [inline lib/atom-css-sniffer.js]: commands

File: lib/atom-css-sniffer.js [hunk 2]
```diff
@@ -4,16 +4,96 @@ let helpers;
...
+    }
+    if (!path) {
+        path = require('path');
+    }
+    if (!filesystem) {
+        filesystem = require('fs');
+    }
+}
+
+function _getCommand(projectPath) {
```
Comment (rickprent) [inline lib/atom-css-sniffer.js]: - project_path
- Do we add JSDoc to functions?

File: lib/atom-css-sniffer.js [hunk 3]
```diff
@@ -4,16 +4,96 @@ let helpers;
...
+    if (cmds[projectPath]) {
+        return cmds[projectPath];
+    }
+
+    let cmd = projectPath + '/vendor/bin/css-sniff';
```
Comment (rickprent) [inline lib/atom-css-sniffer.js]: command_path or path_to_command

File: lib/atom-css-sniffer.js [hunk 4]
```diff
@@ -4,16 +4,96 @@ let helpers;
...
+
+    return cmds[projectPath] = cmd;
+}
+
+function _checkVersion(projectPath, cmd) {
+    if (versions[projectPath]) {
+        return versions[projectPath];
+    }
+
+    return versions[projectPath] = helpers.exec(cmd, ['--version'], {}).then(function (result) {
```
Comment (rickprent) [inline lib/atom-css-sniffer.js]: Is the empty array argument necessary?

File: lib/atom-css-sniffer.js [hunk 5]
```diff
@@ -4,16 +4,96 @@ let helpers;
...
+        let matches = result.match(/v([0-9]\.[0-9])/);
```
Comment (rickprent) [inline lib/atom-css-sniffer.js]: Should we support `1.2.3` and `1.23` ?

File: lib/atom-css-sniffer.js [hunk 6]
```diff
@@ -4,16 +4,96 @@ let helpers;
...
+
+        return matches === null ? '1.0' : matches[1];
+    })
+}
+
+function _processResultV2(data, editorPath) {
+    let errors = [];
+
+    for (let i in data.files) {
```
Comment (rickprent) [inline lib/atom-css-sniffer.js]: I think our JS codestyle defines that we should define all variables of the same scope in one define block, i.e.:

```JS
let errors = [],
    i;
```

**Review by rickprent — CHANGES_REQUESTED** [review]

File: lib/atom-css-sniffer.js (hunk 5 above)
Comment (yannickl88) [reply lib/atom-css-sniffer.js]: no, the version should be API version of the sniffer. so 2.0 is fine.

This value is hard-coded, see https://github.com/hostnet/css-sniffer/blob/1fc0a660fb0f345159ba6bdb7c25b6c581c9f55b/bin/css-sniff#L27

File: lib/atom-css-sniffer.js (hunk 2 above)
Comment (yannickl88) [reply lib/atom-css-sniffer.js]: I think we try to follow the [angular style rules](https://angular.io/guide/styleguide#coding-conventions). However as long as there is no sniffer it's up in the air tbh.

File: lib/atom-css-sniffer.js (hunk 6 above)
Comment (yannickl88) [reply lib/atom-css-sniffer.js]: I'm not sure if I agree that this is more readable...

```js
let errors = [], i;

for (i = 0; i < data.length; i++) {
    // ...
}
```

**Review by linaori — APPROVED** [review]

**Review by rickprent — APPROVED** [review]

**Pull request merged**